provisioning/device/src/transport_state_machine.ts

```diff
@@ -0,0 +1,282 @@
+              }
+            });
+          },
+          disconnect: (callback) => {
+            /* Codes_SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_025: [ If `disconnect` is called while disconnected, it shall immediately call its `callback`. ] */
+            // nothing to do.
+            callback();
+          }
+        },
+        connecting: {
+          _onEnter: (callback) => {
+            debug('entering connecting state');
+            debug('calling transport _doConnectForFsm');
+            /* Codes_SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_002: [ `connect` shall call `_doConnectForFsm`. ] */
+            this._doConnectForFsm((err) => {
+              /* Codes_SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_003: [ If `_doConnectForFsm` fails, then `connect` shall fail. ] */
+              if (err) {
+                /* Codes_SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_011: [ `register` shall fail if the connection fails. ] */
+                this._fsm.transition('disconnecting', callback, err);
+              } else {
+                this._fsm.transition('connected', callback);
+              }
+            });
+          },
+          '*': () => this._fsm.deferUntilTransition()
+        },
+        connected: {
+          _onEnter: (callback, err, body, result) => {
+            debug('entering connected state');
+            callback(err, body, result);
+          },
+          connect: (callback) => {
+            /* Codes_SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_004: [ If the transport is already connected, then `connect` shall do nothing and call the `callback` immediately. ] */
+            // nothing to do.
+            callback();
+          },
+          disconnect: (callback) => {
+            this._fsm.transition('disconnecting', callback);
+          },
+          register: (callback, registrationId, authorization, requestBody, forceRegistration) => {
+            this._fsm.transition('sendingRegistrationRequest', callback, registrationId, authorization, requestBody, forceRegistration);
+          },
+        },
+        sendingRegistrationReqest: {
+          _onEnter: (callback, registrationId, authorization, requestBody, forceRegistration) => {
+            debug('entering sendingRegistrationRequest state');
+            this._registrationCallback = callback;
+            debug('calling transport _doFirstRegistrationRequestForFsm');
+            /* Codes_SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_012: [ `register` shall call `_doFirstRegistrationRequestForFsm`. ] */
+            this._doFirstRegistrationRequestForFsm(registrationId, authorization, requestBody, forceRegistration, (err, responseBody, result, pollingInterval) => {
+              if (this._registrationInProgress()) { // make sure we weren't cancelled before doing something with the response
```

Comment:
>_registrationInProgress [](start = 23, length = 23)

this looks like an inner state within a state machine. Suggest to handle the response with a state machine action, and modify our list of states handle the response based on the state of the state machine.